### PR TITLE
chore: add a dev utility to list issues with instrumention module version ranges

### PR DESCRIPTION
**Pull request by trentm**
'dev-utils/bitrot.js' is a small CLI that will compare the entries in ".tav.yml" and "supported-technologies.asciidoc" for inconsistencies and for whether our supported instrumentation version ranges are behind current releases.

Helpfully, this parsing also noticed a couple errors in our .tav.yml (missing "name: ..." params) that were resulting in use not testing some older module versions.

---

A current run of bitrot shows:

```
% ./dev-utils/bitrot.js
knex bitrot: latest knex@2.3.0 (released 2022-08-31): is not in .tav.yml ranges (^0.16.5 || <0.16.4 >=0.16.0 || ^0.15.0 || ^0.14.0 || ^0.13.0 || ^0.12.5 || <0.12.4 >0.11.6 || <0.11.6 >0.9.0, >=0.17 <0.21, >=0.21 <1), is not in supported-technologies.asciidoc ranges (>=0.9.0 <1.0.0)
mongoose bitrot: latest mongoose@6.6.0 (released 2022-09-08): is not in .tav.yml ranges (), is not in supported-technologies.asciidoc ranges (>=4.0.0 <5.7.0)
redis bitrot: latest redis@4.3.1 (released 2022-09-06): is not in .tav.yml ranges (>=2.0.0 <4.0.0), is not in supported-technologies.asciidoc ranges (>=2.0.0 <4.0.0)
ws bitrot: latest ws@8.8.1 (released 2022-07-15): is not in .tav.yml ranges (>=1 <7, >=7 <8), is not in supported-technologies.asciidoc ranges (>=1.0.0 <8.0.0)
```

We are already on the 'redis' one. We have issues for knex and ws. I have a draft, old, starter PR for mongoose testing -- though priority is probably pretty low (just adding some testing).

### CODE REVIEW

File: docs/supported-technologies.asciidoc [hunk 1]
```diff
@@ -130,23 +127,23 @@ The Node.js agent will automatically instrument the following modules to give yo
...
 |https://www.npmjs.com/package/mongoose[mongoose] |>=4.0.0 <5.7.0 |Supported via mongodb-core
 |https://www.npmjs.com/package/mysql[mysql] |^2.0.0 |Will instrument all queries
 |https://www.npmjs.com/package/mysql2[mysql2] |>=1.0.0 <3.0.0 |Will instrument all queries
-|https://www.npmjs.com/package/pg[pg] |>=4.0.0 <9.0.0 |Will instrument all queries
+|https://www.npmjs.com/package/pg[pg] |>=4.0.0 <9.0.0 |Will instrument all queries
 |https://www.npmjs.com/package/redis[redis] |>=2.0.0 <4.0.0 |Will instrument all queries
 |https://www.npmjs.com/package/tedious[tedious] |>=1.9 <16.0.0 | (Excluding v4.0.0.) Will instrument all queries
 |https://www.npmjs.com/package/undici[undici] | >=4.7.1 <6 | Will instrument undici HTTP requests, except HTTP CONNECT. Requires node v14.17.0 or later, or the user to have installed the https://www.npmjs.com/package/diagnostics_channel['diagnostics_channel' polyfill].
-|https://www.npmjs.com/package/ws[ws] |>=1.0.0 <8.0.0 |Will instrument outgoing WebSocket messages
+|https://www.npmjs.com/package/ws[ws] |>=1.0.0 <8.0.0 |Will instrument outgoing WebSocket messages
```
Comment (trentm) [inline docs/supported-technologies.asciidoc:146]: REVIEW NOTE: These whitespace changes are changing from some weird high-Unicode space char to a vanilla ASCII space.

**Comment by trentm** [comment]
run module tests for graphql,pug,@koa/router

**Comment by trentm** [comment]
> Helpfully, this parsing also noticed a couple errors in our .tav.yml (missing "name: ..." params) that were resulting in use not testing some older module versions.

It turns out that there *is* a breakage lurking in the missed "graphql" tests. The changes for graphql@16 support in #2761 resulted in this breakage with graphql@0.9.6. From https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp/detail/PR-2924/4/pipeline/1066

```
$ .ci/scripts/test.sh -b "release" -t "graphql" "10" 
...
[2022-09-13T19:00:25.328Z] node_tests_1  | -- required packages ["graphql@0.9.6"]
[2022-09-13T19:00:25.328Z] node_tests_1  | -- installing ["graphql@0.9.6"]
[2022-09-13T19:00:33.353Z] node_tests_1  | -- running test "node test/instrumentation/modules/graphql.test.js" with graphql
[2022-09-13T19:00:34.758Z] node_tests_1  | /app/node_modules/graphql/jsutils/invariant.js:19
[2022-09-13T19:00:34.758Z] node_tests_1  |     throw new Error(message);
[2022-09-13T19:00:34.758Z] node_tests_1  |     ^
[2022-09-13T19:00:34.758Z] node_tests_1  | 
[2022-09-13T19:00:34.758Z] node_tests_1  | Error: Must provide document
[2022-09-13T19:00:34.758Z] node_tests_1  |     at invariant (/app/node_modules/graphql/jsutils/invariant.js:19:11)
[2022-09-13T19:00:34.758Z] node_tests_1  |     at Object.execute (/app/node_modules/graphql/execution/execute.js:102:27)
[2022-09-13T19:00:34.758Z] node_tests_1  |     at Object.wrappedExecute (/app/lib/instrumentation/modules/graphql.js:110:21)
[2022-09-13T19:00:34.758Z] node_tests_1  |     at Test.<anonymous> (/app/test/instrumentation/modules/graphql.test.js:182:11)
[2022-09-13T19:00:34.758Z] node_tests_1  |     at Test.bound [as _cb] (/app/node_modules/tape/lib/test.js:99:32)
[2022-09-13T19:00:34.758Z] node_tests_1  |     at Test.run (/app/node_modules/tape/lib/test.js:117:31)
[2022-09-13T19:00:34.758Z] node_tests_1  |     at Test.bound [as run] (/app/node_modules/tape/lib/test.js:99:32)
[2022-09-13T19:00:34.758Z] node_tests_1  |     at Immediate.next (/app/node_modules/tape/lib/results.js:88:19)
[2022-09-13T19:00:34.758Z] node_tests_1  |     at runCallback (timers.js:705:18)
[2022-09-13T19:00:34.758Z] node_tests_1  |     at tryOnImmediate (timers.js:676:5)
[2022-09-13T19:00:34.758Z] node_tests_1  |     at processImmediate (timers.js:658:5)
[2022-09-13T19:00:34.758Z] node_tests_1  | -- detected failing command, flushing stdout...
[2022-09-13T19:00:34.758Z] node_tests_1  | TAP version 13
[2022-09-13T19:00:34.758Z] node_tests_1  | # graphql.graphql(...) - positional args
[2022-09-13T19:00:34.758Z] node_tests_1  | ok 1 no currentSpan in sync code after .graphql(...)
[2022-09-13T19:00:34.758Z] node_tests_1  | ok 2 no currentSpan .graphql().then(...)
[2022-09-13T19:00:34.758Z] node_tests_1  | ok 3 should be loosely deeply equivalent
[2022-09-13T19:00:34.758Z] node_tests_1  | {"log.level":"error","@timestamp":"2022-09-13T19:00:33.918Z","log":{"logger":"elastic-apm-node"},"ecs":{"version":"1.6.0"},"message":"APM Server transport error: error fetching APM Server version: connect ECONNREFUSED 127.0.0.1:8200"}
[2022-09-13T19:00:34.758Z] node_tests_1  | ok 4 should be strictly equal
[2022-09-13T19:00:34.758Z] node_tests_1  | ok 5 should be strictly equal
[2022-09-13T19:00:34.758Z] node_tests_1  | ok 6 should be strictly equal
[2022-09-13T19:00:34.758Z] node_tests_1  | ok 7 should be strictly equal
[2022-09-13T19:00:34.758Z] node_tests_1  | ok 8 should be strictly equal
[2022-09-13T19:00:34.758Z] node_tests_1  | ok 9 should be strictly equal
[2022-09-13T19:00:34.758Z] node_tests_1  | ok 10 should be strictly equal
[2022-09-13T19:00:34.758Z] node_tests_1  | ok 11 should be strictly equal
[2022-09-13T19:00:34.758Z] node_tests_1  | ok 12 should be truthy
[2022-09-13T19:00:34.758Z] node_tests_1  | # graphql.graphql(...) - single GraphQLArgs arg
[2022-09-13T19:00:34.758Z] node_tests_1  | ok 13 no currentSpan in sync code after .graphql(...)
[2022-09-13T19:00:34.758Z] node_tests_1  | ok 14 no currentSpan .graphql().then(...)
[2022-09-13T19:00:34.758Z] node_tests_1  | not ok 15 should be loosely deeply equivalent
[2022-09-13T19:00:34.758Z] node_tests_1  |   ---
[2022-09-13T19:00:34.758Z] node_tests_1  |     operator: deepLooseEqual
[2022-09-13T19:00:34.758Z] node_tests_1  |     expected: |-
[2022-09-13T19:00:34.758Z] node_tests_1  |       { data: { hello: 'Hello world!' } }
[2022-09-13T19:00:34.758Z] node_tests_1  |     actual: |-
[2022-09-13T19:00:34.758Z] node_tests_1  |       { errors: [ GraphQLError { message: 'Syntax Error GraphQL request (1:1) Unexpected <EOF>\n\n1: \n   ^\n', locations: [ { line: 1, column: 1 } ], path: undefined } ] }
[2022-09-13T19:00:34.758Z] node_tests_1  |     at: <anonymous> (/app/test/instrumentation/modules/graphql.test.js:70:7)
[2022-09-13T19:00:34.758Z] node_tests_1  |     stack: |-
[2022-09-13T19:00:34.758Z] node_tests_1  |       Error: should be loosely deeply equivalent
[2022-09-13T19:00:34.759Z] node_tests_1  |           at Test.assert [as _assert] (/app/node_modules/tape/lib/test.js:314:54)
[2022-09-13T19:00:34.759Z] node_tests_1  |           at Test.bound [as _assert] (/app/node_modules/tape/lib/test.js:99:32)
[2022-09-13T19:00:34.759Z] node_tests_1  |           at Test.deepLooseEqual (/app/node_modules/tape/lib/test.js:596:10)
[2022-09-13T19:00:34.759Z] node_tests_1  |           at Test.bound [as deepLooseEqual] (/app/node_modules/tape/lib/test.js:99:32)
[2022-09-13T19:00:34.759Z] node_tests_1  |           at /app/test/instrumentation/modules/graphql.test.js:70:7
[2022-09-13T19:00:34.759Z] node_tests_1  |           at process._tickCallback (internal/process/next_tick.js:68:7)
[2022-09-13T19:00:34.759Z] node_tests_1  |   ...
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 16 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 17 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 18 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 19 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 20 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 21 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 22 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 23 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 24 should be truthy
[2022-09-13T19:00:34.759Z] node_tests_1  | # graphql.graphql - invalid query
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 25 no currentSpan in sync code after .graphql(...)
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 26 no currentSpan .graphql().then(...)
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 27 should be deeply equivalent
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 28 should have one error
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 29 should return a sytax error
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 30 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 31 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 32 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 33 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 34 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 35 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 36 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 37 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 38 should be truthy
[2022-09-13T19:00:34.759Z] node_tests_1  | # graphql.graphql - transaction ended
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 39 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 40 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 41 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 42 should be strictly equal
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 43 no currentSpan in sync code after .graphql(...)
[2022-09-13T19:00:34.759Z] node_tests_1  | ok 44 no currentSpan .graphql().then(...)
[2022-09-13T19:00:34.759Z] node_tests_1  | not ok 45 should be loosely deeply equivalent
[2022-09-13T19:00:34.759Z] node_tests_1  |   ---
[2022-09-13T19:00:34.760Z] node_tests_1  |     operator: deepLooseEqual
[2022-09-13T19:00:34.760Z] node_tests_1  |     expected: |-
[2022-09-13T19:00:34.760Z] node_tests_1  |       { data: { hello: 'Hello world!' } }
[2022-09-13T19:00:34.760Z] node_tests_1  |     actual: |-
[2022-09-13T19:00:34.760Z] node_tests_1  |       { errors: [ GraphQLError { message: 'Syntax Error GraphQL request (1:1) Unexpected <EOF>\n\n1: \n   ^\n', locations: [ { line: 1, column: 1 } ], path: undefined } ] }
[2022-09-13T19:00:34.760Z] node_tests_1  |     at: <anonymous> (/app/test/instrumentation/modules/graphql.test.js:125:7)
[2022-09-13T19:00:34.760Z] node_tests_1  |     stack: |-
[2022-09-13T19:00:34.760Z] node_tests_1  |       Error: should be loosely deeply equivalent
[2022-09-13T19:00:34.760Z] node_tests_1  |           at Test.assert [as _assert] (/app/node_modules/tape/lib/test.js:314:54)
[2022-09-13T19:00:34.760Z] node_tests_1  |           at Test.bound [as _assert] (/app/node_modules/tape/lib/test.js:99:32)
[2022-09-13T19:00:34.760Z] node_tests_1  |           at Test.deepLooseEqual (/app/node_modules/tape/lib/test.js:596:10)
[2022-09-13T19:00:34.760Z] node_tests_1  |           at Test.bound [as deepLooseEqual] (/app/node_modules/tape/lib/test.js:99:32)
[2022-09-13T19:00:34.760Z] node_tests_1  |           at /app/test/instrumentation/modules/graphql.test.js:125:7
[2022-09-13T19:00:34.760Z] node_tests_1  |           at process._tickCallback (internal/process/next_tick.js:68:7)
[2022-09-13T19:00:34.760Z] node_tests_1  |   ...
[2022-09-13T19:00:34.760Z] node_tests_1  | # graphql.execute(...) - positional args
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 46 no currentSpan in sync code after .execute(...)
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 47 no currentSpan .execute().then(...)
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 48 should be loosely deeply equivalent
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 49 should be strictly equal
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 50 should be strictly equal
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 51 should be strictly equal
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 52 should be strictly equal
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 53 should be strictly equal
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 54 should be strictly equal
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 55 should be strictly equal
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 56 should be strictly equal
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 57 should be truthy
[2022-09-13T19:00:34.760Z] node_tests_1  | # graphql.execute - transaction ended
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 58 should be strictly equal
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 59 should be strictly equal
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 60 should be strictly equal
[2022-09-13T19:00:34.760Z] node_tests_1  | ok 61 should be strictly equal
[2022-09-13T19:00:34.760Z] node_tests_1  | 
[2022-09-13T19:00:34.760Z] node_tests_1  | -- fatal: Test exited with code 1
```

Our stated support range is `>=0.7.0 <0.11.0 || >=0.11.1 <17`, so this affects users of graphql 0.7.0 to 0.9.6 inclusive.
`graphql@0.9.6` was released in 2017-05-02T00:22:14.084Z, so we *could* also consider dropping support for those 5+ year old versions of the lib.

**Comment by trentm** [comment]
> It turns out that there _is_ a breakage lurking in the missed "graphql" tests.

This will be fixed in https://github.com/elastic/apm-agent-nodejs/pull/2927  (I'll wait for that to go in before this one.)

File: docs/supported-technologies.asciidoc [hunk 2]
```diff
@@ -87,16 +87,13 @@ These are the frameworks that we officially support:
 
 [options="header"]
 |=======================================================================
```
Comment (trentm) [inline docs/supported-technologies.asciidoc:89]: REVIEW NOTE: aThe changes in this file are all either:
- Formatting changes to be more consistent (which helped the parsing of this file being done in "bitrot.js"), or
- Replacing some weird unicode-space characters of unknown origin with simple ASCII space.

**Comment by trentm** [comment]
The only test failure is the "Integration tests" which can't possibly be affected by this PR. I'm ignoring that failure for now.

**Review by astorm — APPROVED** [review]
Nice!  Robots doing grunt work is always a win.  Approving as is -- and it might be worth mentioning this somewhere in a doc (maybe TESTING.md?) so it's not lost in the `./bin` folder. (I was going to suggest we incorporate this as part of our test suite, but the initial build of `/tmp/apm-agent-nodejs-bitrot.cache.json` takes an (unavoidable) non-trivial amount of time.)

**Pull request merged**